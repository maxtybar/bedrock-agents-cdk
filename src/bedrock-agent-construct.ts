import * as path from 'path';
import {
  aws_lambda as lambda,
  aws_logs as logs,
  aws_iam as iam,
  Duration,
  CustomResource,
  Stack,
  custom_resources,
} from 'aws-cdk-lib';
import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';
import {
  ActionGroup,
  BedrockAgentProps,
  KnowledgeBaseAssociation,
} from './interfaces';

// Assign default props
const defaultProps: Partial<BedrockAgentProps> = {
  foundationModel: 'anthropic.claude-v2',
  idleSessionTTLInSeconds: 3600,
};


// Create a BedrockAgent construct
export class BedrockAgent extends Construct {

  readonly region: string;
  readonly agentName: string;
  readonly instruction: string;
  readonly foundationModel: string;
  readonly agentResourceRoleArn: string;
  readonly description: string | undefined;
  readonly idleSessionTTLInSeconds: number;
  readonly actionGroups: ActionGroup[];
  readonly knowledgeBaseAssociations: KnowledgeBaseAssociation[];
  readonly bedrockAgentCustomResourceRole: iam.Role;

  constructor(scope: Construct, name: string, props: BedrockAgentProps) {
    super(scope, name);

    props = { ...defaultProps, ...props };

    this.region = Stack.of(this).region;
    this.agentName = props.agentName;
    this.instruction = props.instruction;
    this.foundationModel = props.foundationModel as string;
    this.agentResourceRoleArn = props.agentResourceRoleArn ?? this.getDefaultAgentResourceRoleArn();
    this.description = props.description;
    this.idleSessionTTLInSeconds = props.idleSessionTTLInSeconds as number;
    this.actionGroups = props.actionGroups ?? this.getUndefinedActionGroup();
    this.knowledgeBaseAssociations = props.knowledgeBaseAssociations ?? this.getUndefinedKnowledgeBase();

    this.bedrockAgentCustomResourceRole = new iam.Role(this, 'BedrockAgentCustomResourceRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
    });

    // /**
    // * Check if IAM role was provided for knowledge base and if it was
    // * then add iam:PassRole permission on that role to custom resource's
    // * IAM role.
    // */
    // const resources = [this.agentResourceRoleArn];
    // if (this.knowledgeBase.roleArn !== 'Undefined') {
    //   resources.push(this.knowledgeBase.roleArn);
    // }

    this.bedrockAgentCustomResourceRole.addToPolicy(new iam.PolicyStatement({
      actions: ['iam:PassRole'],
      resources: [this.agentResourceRoleArn],
    }));

    this.bedrockAgentCustomResourceRole.addToPolicy(new iam.PolicyStatement({
      actions: ['*'],
      resources: ['arn:aws:bedrock:*'],
    }));

    /**
    * If agent group was provided then attach lambda:AllowInvoke policy
    * by an agent to the provided Lambda.
    */
    // this.actionGroups.actionGroupExecutor != 'Undefined' ? this.lambdaAttachResourceBasedPolicy(): 'Undefined';
    this.actionGroups.forEach(actionGroup => actionGroup.actionGroupExecutor != 'Undefined' ?
      this.lambdaAttachResourceBasedPolicy(actionGroup.actionGroupExecutor) : 'Undefined');

    /**
    * If agent group was provided but IAM role for the agent
    * was not then attach s3 read access to the bucket to the default role.
    */
    // props.actionGroups?.s3BucketName && !props.agentResourceRoleArn && this.attachS3BucketReadOnlyPolicy();
    props.actionGroups?.forEach(actionGroup =>
      actionGroup.s3BucketName && !props.agentResourceRoleArn && this.attachS3BucketReadOnlyPolicy(actionGroup.s3BucketName),
    );

    const layer = new lambda.LayerVersion(this, 'BedrockAgentLayer', {
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/lambda-layer/bedrock-agent-layer.zip')),
      compatibleRuntimes: [lambda.Runtime.PYTHON_3_10],
    });

    const onEvent = new lambda.Function(this, 'BedrockAgentCustomResourceFunction', {
      runtime: lambda.Runtime.PYTHON_3_10,
      handler: 'bedrock_agent_custom_resource.on_event',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/functions')),
      architecture: lambda.Architecture.X86_64,
      layers: [layer],
      timeout: Duration.seconds(600),
      environment: {
        AGENT_NAME: this.agentName,
        INSTRUCTION: this.instruction,
        FOUNDATION_MODEL: this.foundationModel,
        AGENT_RESOURCE_ROLE_ARN: this.agentResourceRoleArn,
        DESCRIPTION: this.description ?? 'Undefined',
        IDLE_SESSION_TTL_IN_SECONDS: this.idleSessionTTLInSeconds.toString(),
        ACTION_GROUPS: JSON.stringify(this.actionGroups),
        KNOWLEDGE_BASE_ASSOCIATIONS: JSON.stringify(this.knowledgeBaseAssociations),
      },
      role: this.bedrockAgentCustomResourceRole,
    });

    const bedrockAgentCustomResourceProvider = new custom_resources.Provider(this, 'BedrockAgentCustomResourceProvider', {
      onEventHandler: onEvent,
      logRetention: logs.RetentionDays.ONE_DAY,
    });

    new CustomResource(this, 'BedrockAgentCustomResource', {
      serviceToken: bedrockAgentCustomResourceProvider.serviceToken,
    });
  }

  private getDefaultAgentResourceRoleArn(): string {
    return new iam.Role(this, 'AgentIamRole', {
      roleName: 'AmazonBedrockExecutionRoleForAgents_' + this.agentName,
      assumedBy: new iam.ServicePrincipal('bedrock.amazonaws.com'),
      description: 'Agent role created by CDK.',
    }).roleArn;
  }

  private attachS3BucketReadOnlyPolicy(s3BucketName: string): void {
    const agentRole = iam.Role.fromRoleArn(this, 'AgentIamRoleArn', this.agentResourceRoleArn);
    const policyStatement =
      new PolicyStatement({
        actions: ['s3:GetObject*', 's3:ListBucket'],
        resources: [`arn:aws:s3:::${s3BucketName}/*`,
          `arn:aws:s3:::${s3BucketName}`],
      });
    agentRole.addToPrincipalPolicy(policyStatement);
  }

  private lambdaAttachResourceBasedPolicy(actionGroupExecutor: string): void {
    this.bedrockAgentCustomResourceRole.addToPolicy(new iam.PolicyStatement({
      actions: ['lambda:AddPermission', 'lambda:GetFunction', 'lambda:RemovePermission'],
      resources: [actionGroupExecutor],
    }));
  }

  private getUndefinedActionGroup(): ActionGroup[] {
    return [{
      actionGroupName: 'Undefined',
      actionGroupExecutor: 'Undefined',
      s3BucketName: 'Undefined',
      s3ObjectKey: 'Undefined',
    }];
  }

  private getUndefinedKnowledgeBase(): KnowledgeBaseAssociation[] {
    return [{
      knowledgeBaseName: 'Undefined',
      instruction: 'Undefined',
    }];
  }
}

