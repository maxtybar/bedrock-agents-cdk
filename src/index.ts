export * from './bedrock-agent-construct';
export * from './interfaces';