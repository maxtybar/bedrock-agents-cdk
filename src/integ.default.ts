import {
  App,
  Stack,
} from 'aws-cdk-lib';
import { BedrockAgent } from './bedrock-agent-construct';

const app = new App();

const stack = new Stack(app, 'BedrockAgentStack');

const agentName = 'MyTestAgent';
const kbName = 'MyTestKnowledgeBase';
const agentRoleArn = 'yourAgentRoleArn';
const kbRoleArn = 'yourKnowledgeBaseRolArn';
const actionGroupName = 'MyTestActionGroup';
const dataSourceName = 'MyDataSource';
const foundationModel = 'anthropic.claude-instant-v1';
const agentInstruction = 'This is a template instruction for my agent. You were created by AWS CDK.';
const kbInstruction = 'This is a template instruction for my knowledge base. You were created by AWS CDK.';
const collectionArn = 'yourOpenSearchServerlessCollectionArn';
const vectorIndexName = 'vector-index-name';
const vectorFieldName = 'vector-field-name';
const textField = 'text-field';
const metadataField = 'metadata-field';
const opensearchStorageConfigurationType = 'OPENSEARCH_SERVERLESS';
const dataSourceType = 'S3';
const dataSourceBucketArn = 'yourDataSourceBucketArn';
const actionGroupLambdaArn = 'yourActionGroupLambdaArn';
const actionGroupS3BucketName = 'yourActionGroupApiSchemaBucketName';
const actionGroupS3ObjectKey = 'yourActionGroupApiSchemaKey';


// Amazon Bedrock Agent and Knowledge Base backed by Opensearch Serverless
new BedrockAgent(stack, 'BedrockAgentAndKbConstruct', {
  agentName: agentName,
  instruction: agentInstruction,
  foundationModel: foundationModel,
  agentResourceRoleArn: agentRoleArn,
  actionGroup: {
    actionGroupName: actionGroupName,
    actionGroupExecutor: actionGroupLambdaArn,
    s3BucketName: actionGroupS3BucketName,
    s3ObjectKey: actionGroupS3ObjectKey,
  },
  knowledgeBase: {
    name: kbName,
    roleArn: kbRoleArn,
    instruction: kbInstruction,
    storageConfiguration: {
      opensearchServerlessConfiguration: {
        collectionArn: collectionArn,
        fieldMapping: {
          metadataField: metadataField,
          textField: textField,
          vectorField: vectorFieldName,
        },
        vectorIndexName: vectorIndexName,
      },
      type: opensearchStorageConfigurationType,
    },
    dataSource: {
      name: dataSourceName,
      dataSourceConfiguration: {
        s3Configuration: {
          bucketArn: dataSourceBucketArn,
        },
        type: dataSourceType,
      },
    },
  },
});
