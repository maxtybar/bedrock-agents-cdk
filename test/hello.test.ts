import {
  App, Stack, assertions,
} from 'aws-cdk-lib';
import { BedrockAgent } from '../src';


test('Match snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'BedrockAgentStack');

  new BedrockAgent(stack, 'BedrockAgentStack', {
    agentName: 'MyAgentTestName',
    instruction: 'This is a template instruction for my agent. You were created by AWS TypeScript CDK.',
    foundationModel: 'anthropic.claude-instant-v1',
    description: 'My custom description for an agent.',
  });

  const template = assertions.Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});

