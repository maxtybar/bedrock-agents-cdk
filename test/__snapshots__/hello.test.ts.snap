// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BedrockAgentStackAgentIamRoleF90DDD6B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "bedrock.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Agent role created by CDK.",
        "RoleName": "AmazonBedrockExecutionRoleForAgents_MyAgentTestName",
      },
      "Type": "AWS::IAM::Role",
    },
    "BedrockAgentStackBedrockAgentCustomResourceA2D15E40": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventFBAAC3C3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "BedrockAgentStackBedrockAgentCustomResourceFunction7C54E154": {
      "DependsOn": [
        "BedrockAgentStackBedrockAgentCustomResourceRoleDefaultPolicyEFD655A7",
        "BedrockAgentStackBedrockAgentCustomResourceRole688B37A5",
      ],
      "Properties": {
        "Architectures": [
          "x86_64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ba878896dd1328e5f6826200c7120c54603686b880d154b3b834922d6735e884.zip",
        },
        "Environment": {
          "Variables": {
            "ACTION_GROUPS": "[{"actionGroupName":"Undefined","actionGroupExecutor":"Undefined","s3BucketName":"Undefined","s3ObjectKey":"Undefined"}]",
            "AGENT_NAME": "MyAgentTestName",
            "AGENT_RESOURCE_ROLE_ARN": {
              "Fn::GetAtt": [
                "BedrockAgentStackAgentIamRoleF90DDD6B",
                "Arn",
              ],
            },
            "DESCRIPTION": "My custom description for an agent.",
            "FOUNDATION_MODEL": "anthropic.claude-instant-v1",
            "IDLE_SESSION_TTL_IN_SECONDS": "3600",
            "INSTRUCTION": "This is a template instruction for my agent. You were created by AWS TypeScript CDK.",
            "KNOWLEDGE_BASE_ASSOCIATIONS": "[{"knowledgeBaseName":"Undefined","instruction":"Undefined"}]",
          },
        },
        "Handler": "bedrock_agent_custom_resource.on_event",
        "Layers": [
          {
            "Ref": "BedrockAgentStackBedrockAgentLayerA4E93A56",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "BedrockAgentStackBedrockAgentCustomResourceRole688B37A5",
            "Arn",
          ],
        },
        "Runtime": "python3.10",
        "Timeout": 600,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventFBAAC3C3": {
      "DependsOn": [
        "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventServiceRoleDefaultPolicy069F488F",
        "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventServiceRoleA42A439E",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7382a0addb9f34974a1ea6c6c9b063882af874828f366f5c93b2b7b64db15c94.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (BedrockAgentStack/BedrockAgentStack/BedrockAgentCustomResourceProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "BedrockAgentStackBedrockAgentCustomResourceFunction7C54E154",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventServiceRoleA42A439E",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventLogRetentionAD594AD8": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventFBAAC3C3",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventServiceRoleA42A439E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventServiceRoleDefaultPolicy069F488F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BedrockAgentStackBedrockAgentCustomResourceFunction7C54E154",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BedrockAgentStackBedrockAgentCustomResourceFunction7C54E154",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventServiceRoleDefaultPolicy069F488F",
        "Roles": [
          {
            "Ref": "BedrockAgentStackBedrockAgentCustomResourceProviderframeworkonEventServiceRoleA42A439E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BedrockAgentStackBedrockAgentCustomResourceRole688B37A5": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BedrockAgentStackBedrockAgentCustomResourceRoleDefaultPolicyEFD655A7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BedrockAgentStackAgentIamRoleF90DDD6B",
                  "Arn",
                ],
              },
            },
            {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "arn:aws:bedrock:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BedrockAgentStackBedrockAgentCustomResourceRoleDefaultPolicyEFD655A7",
        "Roles": [
          {
            "Ref": "BedrockAgentStackBedrockAgentCustomResourceRole688B37A5",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BedrockAgentStackBedrockAgentLayerA4E93A56": {
      "Properties": {
        "CompatibleRuntimes": [
          "python3.10",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5db1ecd01ac13dc1ac81781044759c7d8cc675984d5b6c4d2414e1c6b28aba07.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockAgentCustomResourceFunctionC962F549": {
      "DependsOn": [
        "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceRoleDefaultPolicy90FEEAF3",
        "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceRoleB92EFA5D",
      ],
      "Properties": {
        "Architectures": [
          "x86_64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ba878896dd1328e5f6826200c7120c54603686b880d154b3b834922d6735e884.zip",
        },
        "Environment": {
          "Variables": {
            "DATA_SOURCE": "{"name":"MyDataSource","dataSourceConfiguration":{"s3Configuration":{"bucketArn":"yourDataSourceBucketArn","inclusionPrefixes":["yourDataSourceBucketPrefix"]},"type":"S3"}}",
            "DESCRIPTION": "Undefined",
            "KNOWLEDGEBASE_CONFIGURATION": {
              "Fn::Join": [
                "",
                [
                  "{"type":"VECTOR","vectorKnowledgeBaseConfiguration":{"embeddingModelArn":"arn:aws:bedrock:",
                  {
                    "Ref": "AWS::Region",
                  },
                  "::foundation-model/amazon.titan-embed-text-v1"}}",
                ],
              ],
            },
            "KNOWLEDGEBASE_NAME": "MyTestKnowledgeBase",
            "REMOVAL_POLICY": "destroy",
            "ROLE_ARN": "yourKnowledgeBaseRoleArn",
            "STORAGE_CONFIGURATION": "{"opensearchServerlessConfiguration":{"collectionArn":"yourOpensearchServerlessCollectionArn","fieldMapping":{"metadataField":"metadata-field","textField":"text-field","vectorField":"my-test-vector"},"vectorIndexName":"my-test-index"},"type":"OPENSEARCH_SERVERLESS"}",
          },
        },
        "Handler": "bedrock_knowledgebase_custom_resource.on_event",
        "Layers": [
          {
            "Ref": "BedrockOpenSearchKnowledgeBaseBedrockAgentLayerBF9ED476",
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceRoleB92EFA5D",
            "Arn",
          ],
        },
        "Runtime": "python3.10",
        "Timeout": 600,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockAgentLayerBF9ED476": {
      "Properties": {
        "CompatibleRuntimes": [
          "python3.10",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5db1ecd01ac13dc1ac81781044759c7d8cc675984d5b6c4d2414e1c6b28aba07.zip",
        },
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceF40A0DF4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEvent92B355C1",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEvent92B355C1": {
      "DependsOn": [
        "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventServiceRoleDefaultPolicyACE14FFA",
        "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventServiceRoleBA03F925",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7382a0addb9f34974a1ea6c6c9b063882af874828f366f5c93b2b7b64db15c94.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (BedrockAgentStack/BedrockOpenSearchKnowledgeBase/BedrockKnowledgeBaseCustomResourceProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "BedrockOpenSearchKnowledgeBaseBedrockAgentCustomResourceFunctionC962F549",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventServiceRoleBA03F925",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventLogRetentionBF1B0FC5": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEvent92B355C1",
              },
            ],
          ],
        },
        "RetentionInDays": 1,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventServiceRoleBA03F925": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventServiceRoleDefaultPolicyACE14FFA": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "BedrockOpenSearchKnowledgeBaseBedrockAgentCustomResourceFunctionC962F549",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "BedrockOpenSearchKnowledgeBaseBedrockAgentCustomResourceFunctionC962F549",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventServiceRoleDefaultPolicyACE14FFA",
        "Roles": [
          {
            "Ref": "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceProviderframeworkonEventServiceRoleBA03F925",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceRoleB92EFA5D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceRoleDefaultPolicy90FEEAF3": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": "yourKnowledgeBaseRoleArn",
            },
            {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "arn:aws:bedrock:*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceRoleDefaultPolicy90FEEAF3",
        "Roles": [
          {
            "Ref": "BedrockOpenSearchKnowledgeBaseBedrockKnowledgeBaseCustomResourceRoleB92EFA5D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5bc602ecde93c947efe5899ae355f999986a1acbe610b1c0b9c468d738857555.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
